### ISSUE TITLE
Execution of ReactiveCommand throws sometimes an ObjectDisposedException

### ISSUE THREAD

**Issue by Yarith** [reactiveui/ReactiveUI#1854]
**Do you want to request a *feature* or report a *bug*?**
*bug*

**What is the current behavior?**
Throws an `ObjectDisposedException` if the executing command somehow disposes itself.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
This is a small program which causes the described `ObjectDisposedException`.
```csharp
using ReactiveUI;
using ReactiveUI.Fody.Helpers;
using System;
using System.Diagnostics;
using System.Reactive;
using System.Reactive.Concurrency;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Threading.Tasks;
using System.Windows.Threading;

class Program
{
    static async Task Main(string[] args)
    {
        using (var manager = new CustomerManagerViewModel())
        {
            Debug.Assert(manager.OpenedCustomer == null);
            if (await manager.NewCustomer.CanExecute.FirstOrDefaultAsync())
            {
                await manager.NewCustomer.Execute();
            }
            Debug.Assert(manager.OpenedCustomer != null);
            if (await manager.OpenedCustomer.Leave.CanExecute.FirstOrDefaultAsync())
            {
                await manager.OpenedCustomer.Leave.Execute(); // this line causes ObjectDisposedException
            }
        }
    }
}

class CustomerManagerViewModel : ReactiveObject, IDisposable
{
    private readonly IDisposable serialDisposeLastOpenedCustomer;

    public CustomerManagerViewModel()
    {
        CanLeave = true;

        NewCustomer = ReactiveCommand.Create(() => { },
            outputScheduler: ImmediateScheduler.Instance);
        Subject<Unit> openedCustomerLeft = new Subject<Unit>();
        NewCustomer
            .Select(_ =>
                new CustomerViewModel(
                    canLeaveObservable: this.WhenAnyValue(x => x.CanLeave)))
            .Merge(openedCustomerLeft.Select(_ => (CustomerViewModel)null))
            .ToPropertyEx(this, x => x.OpenedCustomer, scheduler: ImmediateScheduler.Instance);

        // Serial dispose last opened customer        
        serialDisposeLastOpenedCustomer =
            this.WhenAnyValue(x => x.OpenedCustomer)
                .Select(customer =>
                    customer == null ? Observable.Empty<Unit>()
                    : Observable.Create<Unit>(_ => customer))
                .Switch()
                .Subscribe();

        this.WhenAnyObservable(x => x.OpenedCustomer.Leave)
            .Subscribe(openedCustomerLeft);
    }

    public CustomerViewModel OpenedCustomer { [ObservableAsProperty] get; }

    [Reactive]
    public bool CanLeave { get; set; }

    public ReactiveCommand<Unit, Unit> NewCustomer { get; }

    public void Dispose() => serialDisposeLastOpenedCustomer.Dispose();
}

class CustomerViewModel : ReactiveObject, IDisposable
{
    public CustomerViewModel(IObservable<bool> canLeaveObservable)
    {
        if (canLeaveObservable == null)
            throw new ArgumentNullException(nameof(canLeaveObservable));

        Leave = ReactiveCommand.Create(() => { },
            // Register external source as can execute to get a requirement for dispose
            canExecute: canLeaveObservable,
            outputScheduler: ImmediateScheduler.Instance);
    }

    public ReactiveCommand<Unit, Unit> Leave { get; }

    public void Dispose() => Leave.Dispose();
}
```

The disposal of the `_executionInfo`-Subject on the line [ReactiveCommand.cs#L683](https://github.com/reactiveui/ReactiveUI/blob/502963d37c59766c601bf6ac5349c63a461b9eb6/src/ReactiveUI/ReactiveCommand/ReactiveCommand.cs#L683) switches the internal observable to an `DisposedObserver` which throws an `ObjectDisposedException` if something is getting called from it. In this case it is the line [ReactiveCommand.cs#L666](https://github.com/reactiveui/ReactiveUI/blob/502963d37c59766c601bf6ac5349c63a461b9eb6/src/ReactiveUI/ReactiveCommand/ReactiveCommand.cs#L666) which will call the `OnNext` method.

**What is the expected behavior?**
If the execution of the command somehow disposes itself it should not throw on the [finally block](https://github.com/reactiveui/ReactiveUI/blob/502963d37c59766c601bf6ac5349c63a461b9eb6/src/ReactiveUI/ReactiveCommand/ReactiveCommand.cs#L666) of the execution.

**What is the motivation / use case for changing the behavior?**
I have upgraded our program from 7.4.0 to 9.4.1 and got on some places an `ObjectDisposedException`. It gets thrown on the line [ReactiveCommand.cs#L666](https://github.com/reactiveui/ReactiveUI/blob/502963d37c59766c601bf6ac5349c63a461b9eb6/src/ReactiveUI/ReactiveCommand/ReactiveCommand.cs#L666). 

We have many places in our application which has a command that causes the container of the command to get disposed. For example: We have a back button which is bound to a `Leave`-ReactiveCommand this causes the surrounding ViewModel to dispose together with the `Leave`-ReactiveCommand. Also an important information: We use a dispatcher scheduler which will execute the output immediately if it is already on the dispatcher thread. This is the reason why the sample code uses the `ImmediateScheduler`.

**Which versions of ReactiveUI, and which platform / OS are affected by this issue? Did this work in previous versions of ReativeUI? Please also test with the [latest stable](https://www.nuget.org/profiles/reactiveui) and [development snapshot](https://reactiveui.net/docs/getting-started/installation/nuget-packages/development-package-feed)**
This has worked with the version 7.4.0. After the installation of the 9.4.1 version this issue happened. Other versions were not tested.

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
### Possible workaround
I figured out a possible workaround with the observation of the `IsExecuting`-Observable. Together with a `RefCountDisposable` i can prevent the disposal until the execution is completed.
```csharp
SingleAssignmentDisposable isExecutingSubscription = new SingleAssignmentDisposable();
var refCountDisposable = new RefCountDisposable(
    new CompositeDisposable(isExecutingSubscription, reactiveCommand));
disposeAfterExecute = refCountDisposable;
isExecutingSubscription.Disposable =
    reactiveCommand.IsExecuting
        .Select(isExecuting => 
            isExecuting
            ? Observable.Create<Unit>(_ => refCountDisposable.GetDisposable())
            : Observable.Empty<Unit>())
        .Switch()
        .Subscribe();
```
Instead disposing the `reactiveCommand` it will now dispose the `disposeAfterExecute`-Disposable.

### Ideas for a fix
I think this issue was not raised earlier because the `ReactiveCommand` is used most of the time with an deferred scheduler. Maybe the implementation relies on the deferred execution so it passes the finally block before the `Dispose` method gets called. 
1. One way would be to defer the disposal of the `_executionInfo` via an `RefCountDisposable`.
2. Another way would be to mark the `ReactiveCommand` as disposed and do not call the `OnNext` method if it is the case.

Maybe there are even better approaches than these above.

**Comment by benoitjadinon** [comment]
I started getting those right after the rxui 8.x update, still have yet to find a solution, but it's crashing our app a lot :-/

**Comment by glennawatson** [comment]
Usually disposing a ReactiveCommand is not required. It's owned by the current object anyway. 

See here when it is and is not a good idea to dispose. 

https://github.com/cabauman/Rx.Net-ReactiveUI-CheatSheet/blob/master/README.md

**Comment by glennawatson** [comment]
I think the example @yarith has is one example where you may since the can execute observable is passed in from outside the class. Most people create the ohservable from within the same class so the lifetime of the event/delegate is the same so the ReactiveCommand so the GC dependency graph knows it can clear the memory so the above guide may unblock you @benoitjadion

**Comment by benoitjadinon** [comment]
thanks @glennawatson, nice tips.
I just found our issue is happening because we're disposing a command created within a WhenActivated from a CreateFromObservable() on an observable that is never completing. so it probably leaks and triggers a onNext after the command is disposed. makes very much sense now, but the rx stacktraces are rarely helping.
thanks

**Comment by metju30** [comment]
Can anyone solve this? .. I need to call dispose on my commands and I am getting this issue
Or .. how can I(if possible) implement mentioned workaround in my extended ReactiveCommand? 
THX!

**Comment by glennawatson** [comment]
Why do you need to dispose your command? You almost never should.

**Comment by metju30** [comment]
Yes I've seen your comment and I have that situation .. due to a structure of program I have cross referenced objects: 
Command owner is not the same class as execute and canExecute observables. So I'd like to clean up after I'm done using particular instance.
Also it seems pretty elementary functionality and since it has Dispose method I think its important to be working correctly, even if its ALMOST never required.
Its also weird that (in my case) exception only happens when NOT in debugging mode (I guess due to speed of code execution)

Thx for your response!

**Comment by LouisSimon73** [comment]
We have the same problem here: we got a lot of ReactiveCommand everywhere and, as it was IDisposable, we did dispose them everytime. Also some of them do subscribe to services that might be shared in the application... If we must not dispose ReactiveCommand, why do they implement IDisposable ? Generally I think that Disposing should be harmless. Here it works if we dispose all subscriptions to this command and then the command. But this is heavy. 
Thx for any answer on that.

**Comment by glennawatson** [comment]
Don't dispose ReactiveCommand is the general advice, the reality is a lot of IObservable can be not Disposed, it's essentially the wrong pattern in some ways since it's not your traditional meaning of "IDispose" where you should always dispose.

**Comment by LouisSimon73** [comment]
Ok we got the general advise on that and we read it a couple of time already but this is a regression since ReactiveCommands were disposable before. We have hundreds of commands and some MUST be disposed beacause of not we have memory leaks. What do we do with those ?

**Comment by glennawatson** [comment]
In general you just won't have memory leaks is more the point of what conversation. 

I have pushed a change d24e69f2dd47729045507d7f16ad0bb418c6925e which reverts some internal disposing. We only need to dispose the external subscription passed in for the `can change`. I'll release soon, let me know if it helps.

**Comment by glennawatson** [comment]
https://github.com/reactiveui/ReactiveUI/releases/tag/18.2.5 is the release.

**Comment by LouisSimon73** [comment]
Thanks a lot for your work and reactivity !
